exercises-cpp/mohammed_almakki/ex02_oo_basics/src/ShapeSorter.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include "Shape.h"
+
+class ShapeSorter
+{
+public:
+    ShapeSorter(const std::vector<std::shared_ptr<Shape>> &shapes) : shapes{shapes} {}
+
+    void printShapesWithType(const std::string &type)
+    {
+        for (auto shape : shapes)
+        {
+            if (shape->getType() == type)
+            {
+                std::cout << shape->toString() << " ";
+            }
+        }
+        std::cout << "\n";
+    }
+
+    void printShapesWithSides(const int &sidesCount)
+    {
+        for (auto shape : shapes)
```

Comment:
Likely needless complication in this instance, but could be interesting to see if you could reduce code duplication (use of function pointers?)

Reply:
Yes, we can make a general function that prints shapes that meet a specific condition. The function will accept a lambda filter function as a parameter. The lambda function will accept a parameter of type shape and return a Boolean value. Also, we can get rid of the for loop and use for_each and pass the function to it.